# Implementing XGBoost using Grid Search as well as Hyperopt
- modeltype: **XGBoost** 
- train: **0.886**
- test: **0.865**

## Flow of the notebook
* 0. Rationale behind choosing XGBoost
* 1. Reading Data and importing libraries
* 2. Data Manipulation
* 3. Implementing XGBoost

## 0. Rationale behind choosing XGBoost

There are many articles online that has compared XGBoost, CatBoost and LightGBM and in most of them, XGBoost had a relatively lesser efficiency, and in few of them, it also had a relatively lesser accuracy.

But after researching a bit, XGBoost seemed a highly popular algorithm in data scienstist community and also several top performing kagglers had mentioned this algorithm in the past. So even though being a bit older and relatively inefficient compared to CatBoost and LighGBM I was curious to get my hands on it.

I have explained the working and parameter tuning of the algorithm in the section 3.

## 1. Reading Data & Importing Libraries

In [1]:
# Basic Data operations
import os
import numpy as np
import pandas as pd
import pickle


# H20 and Grid Search
import h2o
from h2o.estimators import H2OXGBoostEstimator
from h2o.grid.grid_search import H2OGridSearch


# Hyper Opt
import hyperopt
from numpy.random import RandomState
from sklearn.metrics import roc_auc_score

# Uncomment this if the category_encoder version is less than 2.1.0
# ! pip install category_encoders # This will update Sk-learn as well
from category_encoders import *

In [2]:
# Set directories
print(os.getcwd())
dirRawData = "../input/"
dirPData   = "../PData/"
dirPOutput = "../POutput/"

/home/jovyan/Final_Project/Pcode


In [3]:
f_name = dirPData + '01_df_250k.pickle'

with (open(f_name, "rb")) as f:
    dict_ = pickle.load(f)

df_train = dict_['df_train']
df_test  = dict_['df_test']

del f_name, dict_

In [4]:
f_name = dirPData + '01_vars.pickle'

with open(f_name, "rb") as f:
    dict_ = pickle.load(f)

vars_ind_numeric     = dict_['vars_ind_numeric']
vars_ind_hccv        = dict_['vars_ind_hccv']
vars_ind_categorical = dict_['vars_ind_categorical']
vars_notToUse        = ['unique_id']
var_dep              = dict_['var_dep']

del f_name, dict_

## 2. Data manipulation

### 2.2 Treating NAs

* **Handling NAs**

One of the advantage of XGBoost over other Gradient Boosting algorithm is it efficiently handles the missing values. As Tianqi Chen has discussed in his paper
> When a value is missing in the sparse matrix x, the instance is classified into the default direction. There are two choices
of default direction in each branch. The optimal default directions are learnt from the data. The key improvement is to only visit the non-missing entries. The presented algorithm treats the non-presence as a missing value and learns the best direction to handle missing values. 

And as mentioned in the H20 documentation: 
>  Missing values are interpreted as containing information (i.e., missing for a reason), rather than missing at random. During tree building, split decisions for every node are found by minimizing the loss function and treating missing values as a separate category that can go either left or right. XGBoost will automatically learn which is the best direction to go when a value is missing.

Following this, we have decided not to do any operations on missing value (except for deleting 'c02' column becuase it contain >50% of missing values in test data). 

-99 values will be treated in the H20 Frame, right now we are only removing 'C02' variable.

In [5]:


# Since C02 has a lot of NAs (>50%), we will drop it from both train and test
df_test.drop('c02', axis=1, inplace = True)
df_train.drop('c02', axis=1, inplace = True)
vars_ind_categorical.remove('c02')


### 2.3 Cardinality
   For treating the hccv we have used Target Encoders (Sk-Learn). We have also observed that there is oversampling of few factors 
    while others are under-sampled therefore we have used smoothing factor = 4. Smoothing of 4 is chosen by following few online 
    blogs. Also scikit learn is used to do the target encoding because I was a bit confused with H20 target encoding. Additionally,
    H20 automatically perform one_hot encoding on categorical variables so they have not been treated.


Note: I needed to update the scikit-learn and category_ecoders library

In [8]:
# To see the distribution of factors in a hcc variable
# df_train[vars_ind_hccv].nunique()
# df_train['e18'].value_counts()

In [6]:
# Target encoders on Train
enc = TargetEncoder(cols=vars_ind_hccv, smoothing =4)
enc.fit_transform(df_train, df_train['target'])
df_train = enc.transform(df_train, df_train['target'])

In [7]:
# Target encoders on Test
df_test['target'] = np.nan # Creating dummy 'target' variable for using the the enc.transform function
df_test = enc.transform(df_test) # applying the already trained encoder
df_test.drop(columns=['target'], inplace=True) # Dropping dummy 'target' variable

Running H20 cluster

In [11]:
# h2o.cluster().shutdown()

AttributeError: 'NoneType' object has no attribute 'shutdown'

In [8]:
h2o.init(port=54321, max_mem_size = "14g") # Asking h20 to use 14 GB of Ram
h2o.connect()

Checking whether there is an H2O instance running at http://localhost:54321 ..... not found.
Attempting to start a local H2O server...
  Java Version: openjdk version "1.8.0_212"; OpenJDK Runtime Environment (build 1.8.0_212-8u212-b03-0ubuntu1.18.04.1-b03); OpenJDK 64-Bit Server VM (build 25.212-b03, mixed mode)
  Starting server from /opt/conda/lib/python3.6/site-packages/h2o/backend/bin/h2o.jar
  Ice root: /tmp/tmp17mtc_0v
  JVM stdout: /tmp/tmp17mtc_0v/h2o_jovyan_started_from_python.out
  JVM stderr: /tmp/tmp17mtc_0v/h2o_jovyan_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321 ... successful.


--------------------------  ---------------------------------------------------
H2O cluster uptime:         01 secs
H2O cluster timezone:       Etc/UTC
H2O data parsing timezone:  UTC
H2O cluster version:        3.24.0.3
H2O cluster version age:    1 year, 2 months and 8 days !!!
H2O cluster name:           H2O_from_python_jovyan_s0rtq1
H2O cluster total nodes:    1
H2O cluster free memory:    12.44 Gb
H2O cluster total cores:    4
H2O cluster allowed cores:  4
H2O cluster status:         accepting new members, healthy
H2O connection url:         http://127.0.0.1:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         Amazon S3, XGBoost, Algos, AutoML, Core V3, Core V4
Python version:             3.6.7 final
--------------------------  ---------------------------------------------------

Connecting to H2O server at http://localhost:54321 ... successful.


--------------------------  ---------------------------------------------------
H2O cluster uptime:         02 secs
H2O cluster timezone:       Etc/UTC
H2O data parsing timezone:  UTC
H2O cluster version:        3.24.0.3
H2O cluster version age:    1 year, 2 months and 8 days !!!
H2O cluster name:           H2O_from_python_jovyan_s0rtq1
H2O cluster total nodes:    1
H2O cluster free memory:    12.44 Gb
H2O cluster total cores:    4
H2O cluster allowed cores:  4
H2O cluster status:         locked, healthy
H2O connection url:         http://localhost:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         Amazon S3, XGBoost, Algos, AutoML, Core V3, Core V4
Python version:             3.6.7 final
--------------------------  ---------------------------------------------------

<H2OConnection to http://localhost:54321, no session>

We have removed 'unique_id' from Frames as it was not useful in prediction


In [9]:
vars_to_use = vars_ind_numeric + vars_ind_categorical
vars_to_use.remove('unique_id')
vars_ind_numeric.remove('unique_id')


h2o_df_train = h2o.H2OFrame(df_train[[var for var in vars_to_use+var_dep ]], destination_frame = 'df_train') # Train Frame
h2o_df_test  = h2o.H2OFrame(df_test[[var for var in vars_to_use]], destination_frame = 'df_test') # Test Frame

Parse progress: |█████████████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%


Converting -99 to NAs. We have not done this before becuase as discussed earlier, converting pandas NAs to H20 Frame were not consistent and was giving an error (may be due to the older version of H20 in the image).

In [10]:
# Converting -99 to NA in train
for var in vars_ind_numeric:
    h2o_df_train[h2o_df_train[var] == -99.0 , var] = None
    
# Converting -99 to NA in test
for var in vars_ind_numeric:
    h2o_df_test[h2o_df_test[var] == -99.0 , var] = None

In [11]:
# H20 Document suggest to make dependant variable as factor for classiication task
h2o_df_train[var_dep] = h2o_df_train[var_dep].asfactor()

### 2.4 Interactions
Documentation regarding the interaction in XGBoost was not clear, therefore after  searching for online, especially [here](https://www.kaggle.com/c/bosch-production-line-performance/discussion/24418) making interaction seemed to be a logical choice given it has drastically improved the past model (GLM) performance.

I have not used the interaction on all variables, instead I have used it only on 2 variables ('f03' and 'e11'). These two 
variables have been chosen after running GLM model and then calculating variables importance.
I did not use all categorical variables, as it could not improve the performance significantly but made the
model quite complex to comprehend. Therefore, it was a trade-off between accuracy and complexity.


In [44]:
# Code for runing GLM to see which all variables are important

# Idea has been taken from: https://aichamp.wordpress.com/2017/09/29/python-example-of-building-glm-gbm-and-random-forest-
# binomial-model-with-h2o/


# from h2o.estimators.glm import H2OGeneralizedLinearEstimator
# glm_logistic = H2OGeneralizedLinearEstimator(family = "binomial")
# glm_logistic.train(x=vars_to_use , y= 'target', training_frame=h2o_df_train, model_id="glm_logistic")
# preds = glm_logistic.predict(h2o_df_test)
# df_test['Predicted'] = np.round(preds[2].as_data_frame(), 5)
# df_preds_dt = df_test[['unique_id', 'Predicted']].copy()
# df_test[['unique_id', 'Predicted']].to_csv(dirPOutput + '1st.csv', index=False)
# log_var_imp = glm_logistic.varimp(use_pandas=True).head()
# log_var_imp.loc[0:5, 'variable'].tolist()

glm Model Build progress: |███████████████████████████████████████████████| 100%
glm prediction progress: |████████████████████████████████████████████████| 100%


/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'e20' has levels not trained on: [30146, BE271, nan]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'e13' has levels not trained on: [Q, S]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'e24' has levels not trained on: [J, M, nan]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'e03' has levels not trained on: [J, nan]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'c09' has levels not trained on: [nan]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'a18' has levels not trained on: [D]
  warnings.warn(w)


['f10', 'f03.F', 'e19', 'e11.A', 'f03.E']

I have chosen min_occurence as int(len(h2o_df_train)/40) after many trial and error.

With int(len(h2o_df_train)/40) on 250K train data I was getting 5 factors for each variable. I believe performing 
interactions on top 4-5 variables rather than all the variables having occurrence > 10/20 would make the model
faster and more interpretable

In [12]:
# Train Frame
interaction_frame_train = h2o_df_train.interaction(['f03', 'e11'], pairwise = False, max_factors = 100,
                                                   min_occurrence = int(len(h2o_df_train)/40))

# Test Frame
interaction_frame_test = h2o_df_test.interaction(['f03', 'e11'], pairwise = False, max_factors = 100, 
                                                 min_occurrence = int(len(h2o_df_train)/40))

# Cbinding interaction frame to train and test
h2o_df_train = h2o_df_train.cbind(interaction_frame_train)
h2o_df_test = h2o_df_test.cbind(interaction_frame_test)

Interactions progress: |██████████████████████████████████████████████████| 100%
Interactions progress: |██████████████████████████████████████████████████| 100%


In [13]:
# incuding interaction frame's variable to variables list
vars_to_use = vars_to_use + ['f03_e11']
vars_ind_numeric = vars_ind_numeric + ['f03_e11']

## 3. Implementing XGBoost
* **My understanding of the algorithm**

XGBoost (Extreme Gradient Boosting) is a boosting algorithm. Boosting refers to ensemble technique where several models (tree models) are being used to make the final model.

Boosting models are build sequentially by minimising the errors from the previous model and Gradient Boosting is an advance version of boosting where the algorithm employs gradient descent method to minimise the error. The main difference between Bagging and Boosting is, Bagging combines different models while Boosting use the previous model to make a new model with less errors.

* **One_hot encoding**

Since XGBoost is dependent on Tree based algorithm, it was safe to assume that XGBoost do not require one_hot encoding. But after going through the paper [Chen T.,Guestrin C., XGBoost: A Scalable Tree Boosting System](https://arxiv.org/pdf/1603.02754.pdf) I realised that one_hot encoding is important if there are categorical variables to make XGBoost perfom better. 

After Going through H20 documentation, I realised that the algorithm, similar to GLM, automatically performs one_hot encoding for XGBoost and thus we do not have to manually do that.

* **Spline**

Since XGBoost is based on Tree based algorithm, we have not used splines.

### 3.1 **Grid Search Vs HyperOpt**

For my own understanding, I tried both, Grid Search and HyperOpt.

Grid search processed only 2 algorithm in 2 hours optimising 5 hyper-parameters with 5 nfolds. It's Train AUC was = 0.889 and valid AUC = 0.879. Prediction accuracy on Kaggle for Grid Search algorithm was 0.865.

On the other hand, HyperOpt explored 10 algorithms in 5 hours minutes (limit of 30 mins per algorithm) and returned train AUC of 0.886, validation AUC of 0.878 and Kaggle Test data AUC of 0.865. Even though the accuracy was not actually significant, I still decided to go for Hyperopt to get my hands on it.

### 3.2 Tuning Hyper-Parameters
**Hyper-paramters to optimize**
* Max_depth: It sets the number of maximum tree depths. Increasing the depth makes the model complex and may lead to overfitting. Default value is 6. By going through few tutorials, I have kept the values from 4 to 8.
* sample_rate: sample_rate tell XGBoost how much data sample should be collected to grow the tree. default is 1, which means 100% data is to be used. Higher values may increase the train accuracy, but for the accuracy on test data, sampling is suggested [Friedman, 1999](https://statweb.stanford.edu/~jhf/ftp/stobst.pdf). I have the values of 0.4, 0.6, 0.8 and 0.9
* min_rows: Specify the minimum number of observations for a leaf. Default value is 1, but in our model, value 1 may make the algorithm time consuming and might not significantly improve the accuracy. Therefore, I have chosen value of 10,20,30 and 40
* ntrees: It tell XGBoost the numbers of tree to build. I have tried 200 trees and observed that there is a need to build more trees to get it converged. Also, since we have taken small learn_rate values, we would require more tree. Therefore, I have set the values to 1000, 2000, 3000 and 4000.
* learn_rate: As per my understanding, learn_rate here is similar to learning-rate in gradient boosting. I couldn't get much information from the H20 documentation. Therefore, I reviewed certain online blogs to see how others have comprehended and used this value. After researching for a while I selected 0.01, 0.005 and 0.001 because others have used these values in line to the number of trees we are using.

**Fixed Hyper-parameters** have been discussed in the code.

### Grid Search

In [87]:
max_depth = [4, 5, 6, 7, 8]
sample_rate = [0.4, 0.6, 0.8, 0.9]
min_rows = [10, 20, 30, 40] # 
ntrees = [1000, 2000, 3000, 4000]
learn_rate = [0.01, 0.005, 0.001]

search_crit = {'strategy': "RandomDiscrete",#  The default strategy, "Cartesian", covers the entire space of h-p 
               # combinations but takes too much time. Therefore we have used "Random Discrete Strategy"
               "max_runtime_secs": 7200, #  maximum time allowed for he algorithm to run
               "max_models": 10, # maximum models to return
               "seed": 2020,  
               "stopping_rounds" : 4, # Stops training when the option selected for stopping_metric doesn’t improve for
               # the specified number of training rounds, based on a simple moving average. 
               "stopping_metric" : "AUC",# Since the kaggle score is based upon AUC, therefore I am using AUC as stopping metric
               "stopping_tolerance" : 1e-3} # to stop training if the improvement is less than this value. 1e-3 is the default
                # value

hyper_params = {
      "ntrees" : ntrees 
    , "max_depth" : max_depth 
    , "learn_rate" : learn_rate
    , "sample_rate" : sample_rate
    , "min_rows" : min_rows 
}
XGB_grid = H2OGridSearch(H2OXGBoostEstimator(col_sample_rate_per_tree = .9, # Specify the column subsampling rate per tree. It
                                             # is multiplicative with col_sample_rate, so setting both parameters to 0.8,
                                             # for example, results in 64% of columns being considered at any given node to 
                                             # split.0.9 has been taken because it seemed the most common choice among the 
                                             # online blogs
                                             score_tree_interval = 100, # Scoring the model after every so many trees
                                             nfolds = 5 # nfolds ideally took more time but I believe it is the correct
                                             # way of doing rather than taking validation set.
                                            ),
                         hyper_params=hyper_params,
                         search_criteria=search_crit,
                         grid_id='g2') # name of the grid

XGB_grid.train(x=vars_to_use,
               y='target',
               training_frame=h2o_df_train) 

In [89]:
# exploring the modesl 
XGB_grid.get_grid()

    learn_rate max_depth min_rows ntrees sample_rate   model_ids  accuracy
0        0.005         8     10.0   1360         0.8  g2_model_1  0.782824
1        0.001         8     20.0     46         0.6  g2_model_2  0.773228


In [122]:
# Best model
best_XG_Grid 

Model Details
H2OXGBoostEstimator :  XGBoost
Model Key:  g2_model_1


ModelMetricsBinomial: xgboost
** Reported on train data. **

MSE: 0.13601975447423742
RMSE: 0.3688085607388167
LogLoss: 0.4166825112634517
Mean Per-Class Error: 0.2001375184104801
AUC: 0.8892462043527933
pr_auc: 0.8805401159906923
Gini: 0.7784924087055867
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.41581707225976355: 


0       1       Error    Rate
-----  ------  ------  -------  ------------------
0      90664   36505   0.2871   (36505.0/127169.0)
1      15199   107632  0.1237   (15199.0/122831.0)
Total  105863  144137  0.2068   (51704.0/250000.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.415817     0.806329  232
max f2                       0.188916     0.882462  316
max f0point5                 0.629473     0.809271  152
max accuracy                 0.508725     0.799644  198
max precision                0.991052     1         0
max recall                   0.0077728    1         395
max specificity              0.991052     1         0
max absolute_mcc             0.48895      0.599857  205
max min_per_class_accuracy   0.508725     0.799415  198
max mean_per_class_accuracy  0.48895      0.799862  205

Gains/Lift Table: Avg response rate: 49.13 %, avg score: 49.13 %



group    cumulative_data_fraction    lower_threshold    lift        cumulative_lift    response_rate    score      cumulative_response_rate    cumulative_score    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  ----------  -----------------  ---------------  ---------  --------------------------  ------------------  --------------  -------------------------  --------  -----------------
    1        0.01                        0.980539           2.03369     2.03369            0.9992           0.98613    0.9992                      0.98613             0.0203369       0.0203369                  103.369   103.369
    2        0.02                        0.96947            2.01903     2.02636            0.992            0.974857   0.9956                      0.980493            0.0201903       0.0405272                  101.903   102.636
    3        0.03                        0.961496           2.01985     2.02419            0.9924           0.965242   0.994533                    0.97541             0.0201985       0.0607257                  101.985   102.419
    4        0.04                        0.955231           2.01252     2.02127            0.9888           0.958307   0.9931                      0.971134            0.0201252       0.0808509                  101.252   102.127
    5        0.05                        0.949446           1.9995      2.01692            0.9824           0.95233    0.99096                     0.967373            0.019995        0.100846                   99.9495   101.692
    6        0.1                         0.920674           1.96823     1.99258            0.96704          0.935284   0.979                       0.951328            0.0984116       0.199258                   96.8233   99.2575
    7        0.15                        0.882862           1.88845     1.95787            0.92784          0.902508   0.961947                    0.935055            0.0944224       0.29368                    88.8448   95.7866
    8        0.2                         0.840248           1.8085      1.92052            0.88856          0.861782   0.9436                      0.916737            0.0904251       0.384105                   80.8501   92.0525
    9        0.3                         0.736948           1.64861     1.82989            0.81             0.790276   0.899067                    0.874583            0.164861        0.548966                   64.8607   82.9886
    10       0.4                         0.610311           1.39606     1.72143            0.68592          0.673937   0.84578                     0.824422            0.139606        0.688572                   39.6065   72.143
    11       0.5                         0.501973           1.16298     1.60974            0.5714           0.554187   0.790904                    0.770375            0.116298        0.80487                    16.298    60.974
    12       0.6                         0.387897           0.899366    1.49134            0.44188          0.445164   0.732733                    0.716173            0.0899366       0.894807                   -10.0634  49.1344
    13       0.7                         0.262625           0.604408    1.36464            0.29696          0.327565   0.67048                     0.660658            0.0604408       0.955247                   -39.5592  36.4639
    14       0.8                         0.135081           0.334525    1.23587            0.16436          0.196073   0.607215                    0.602584            0.0334525       0.9887                     -66.5475  23.5875
    15       0.9                         0.030962           0.106081    1.11034            0.05212          0.080358   0.545538                    0.544559            0.0106081       0.999308                   -89.3919  11.0342
    16       1                           0.00146787         0.00692008  1                  0.0034           0.01



ModelMetricsBinomial: xgboost
** Reported on cross-validation data. **

MSE: 0.1419487460194194
RMSE: 0.3767608605195335
LogLoss: 0.433612396356896
Mean Per-Class Error: 0.20900672194385894
AUC: 0.8787883643498352
pr_auc: 0.8688422794421169
Gini: 0.7575767286996704
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.4055378807417919: 


0       1       Error    Rate
-----  ------  ------  -------  ------------------
0      88100   39069   0.3072   (39069.0/127169.0)
1      15225   107606  0.124    (15225.0/122831.0)
Total  103325  146675  0.2172   (54294.0/250000.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.405538     0.798543  240
max f2                       0.19185      0.877633  320
max f0point5                 0.620411     0.798666  156
max accuracy                 0.491313     0.79068   206
max precision                0.992344     1         0
max recall                   0.00355994   1         399
max specificity              0.992344     1         0
max absolute_mcc             0.482697     0.582309  209
max min_per_class_accuracy   0.506087     0.789634  200
max mean_per_class_accuracy  0.482697     0.790993  209

Gains/Lift Table: Avg response rate: 49.13 %, avg score: 49.13 %



group    cumulative_data_fraction    lower_threshold    lift      cumulative_lift    response_rate    score      cumulative_response_rate    cumulative_score    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  --------  -----------------  ---------------  ---------  --------------------------  ------------------  --------------  -------------------------  --------  -----------------
    1        0.01                        0.979646           2.02148   2.02148            0.9932           0.985585   0.9932                      0.985585            0.0202148       0.0202148                  102.148   102.148
    2        0.02                        0.968848           2.01089   2.01618            0.988            0.974105   0.9906                      0.979845            0.0201089       0.0403237                  101.089   101.618
    3        0.03                        0.960928           2.00275   2.01171            0.984            0.964694   0.9884                      0.974795            0.0200275       0.0603512                  100.275   101.171
    4        0.04                        0.954415           1.98891   2.00601            0.9772           0.957653   0.9856                      0.970509            0.0198891       0.0802403                  98.8912   100.601
    5        0.05                        0.948385           1.98728   2.00226            0.9764           0.951445   0.98376                     0.966696            0.0198728       0.100113                   98.7283   100.226
    6        0.1                         0.919071           1.9373    1.96978            0.95184          0.934054   0.9678                      0.950375            0.0968648       0.196978                   93.7296   96.978
    7        0.15                        0.881512           1.86142   1.93366            0.91456          0.900967   0.950053                    0.933906            0.093071        0.290049                   86.1419   93.366
    8        0.2                         0.838002           1.77545   1.89411            0.87232          0.860159   0.93062                     0.915469            0.0887724       0.378821                   77.5448   89.4107
    9        0.3                         0.736009           1.62744   1.80522            0.7996           0.788559   0.886947                    0.873166            0.162744        0.541565                   62.7439   80.5217
    10       0.4                         0.608795           1.38019   1.69896            0.67812          0.673124   0.83474                     0.823155            0.138019        0.679584                   38.0189   69.896
    11       0.5                         0.500504           1.16086   1.59134            0.57036          0.552949   0.781864                    0.769114            0.116086        0.79567                    16.0863   59.1341
    12       0.6                         0.389146           0.906367  1.47718            0.44532          0.445252   0.725773                    0.715137            0.0906367       0.886307                   -9.36327  47.7179
    13       0.7                         0.264424           0.625494  1.35551            0.30732          0.329375   0.665994                    0.660028            0.0625494       0.948857                   -37.4506  35.5509
    14       0.8                         0.136805           0.360495  1.23113            0.17712          0.198103   0.604885                    0.602288            0.0360495       0.984906                   -63.9505  23.1133
    15       0.9                         0.0323017          0.132784  1.10909            0.06524          0.0820556  0.544924                    0.544484            0.0132784       0.998184                   -86.7216  10.9094
    16       1                           0.000934468        0.018155  1                  0.00892          0.0129779  0.491324                    0.


Cross-Validation Metrics Summary: 


mean      sd          cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid
-----------------------  --------  ----------  ------------  ------------  ------------  ------------  ------------
accuracy                 0.783403  0.0027226   0.777556      0.782246      0.784879      0.782914      0.789418
auc                      0.878804  0.00167602  0.87737       0.876517      0.877491      0.879529      0.883111
err                      0.216597  0.0027226   0.222444      0.217754      0.215121      0.217086      0.210582
err_count                10830     142.373     11130         10879         10804         10838         10499
f0point5                 0.75906   0.00383577  0.750583      0.758276      0.763542      0.756805      0.766094
f1                       0.798922  0.00154562  0.798282      0.795126      0.799316      0.801553      0.800335
f2                       0.843302  0.00517518  0.852455      0.835742      0.838608      0.851925      0.837779
lift_top_group           2.0216    0.0111948   2.00421       2.03526       2.00772       2.01598       2.04484
logloss                  0.433605  0.00291833  0.435676      0.437135      0.436716      0.432474      0.426025
max_per_class_error      0.305937  0.0133243   0.334542      0.296948      0.295352      0.320439      0.282401
mcc                      0.57869   0.00373345  0.572212      0.574513      0.578682      0.58062       0.587422
mean_per_class_accuracy  0.784932  0.00272309  0.779153      0.784128      0.785835      0.784369      0.791175
mean_per_class_error     0.215068  0.00272309  0.220847      0.215872      0.214165      0.215631      0.208825
mse                      0.141947  0.00101831  0.142716      0.143369      0.142759      0.141579      0.139309
precision                0.734659  0.00583148  0.721829      0.735549      0.74142       0.729649      0.74485
r2                       0.432028  0.00404552  0.429022      0.426213      0.428885      0.433574      0.442447
recall                   0.875801  0.00883744  0.892848      0.865205      0.867021      0.889178      0.864752
rmse                     0.376753  0.00135479  0.377778      0.378642      0.377835      0.37627       0.373241
specificity              0.694064  0.0133243   0.665458      0.703052      0.704648      0.679561      0.717599

Scoring History: 


timestamp            duration     number_of_trees    training_rmse    training_logloss    training_auc    training_pr_auc    training_lift    training_classification_error
--  -------------------  -----------  -----------------  ---------------  ------------------  --------------  -----------------  ---------------  -------------------------------
    2020-07-11 20:16:51  1:38:02.233  0                  0.5              0.693147            0.5             0                  1                0.508676
    2020-07-11 20:18:05  1:39:15.936  100                0.434947         0.56802             0.871847        0.861966           2.01171          0.225152
    2020-07-11 20:19:15  1:40:26.730  200                0.404            0.506355            0.874372        0.865455           2.01985          0.2203
    2020-07-11 20:20:28  1:41:39.228  300                0.389696         0.473633            0.876379        0.867767           2.02148          0.21852
    2020-07-11 20:21:42  1:42:53.173  400                0.382936         0.455557            0.87802         0.869998           2.02473          0.218492
    2020-07-11 20:22:57  1:44:08.310  500                0.379442         0.444894            0.879247        0.870534           2.02718          0.214904
    2020-07-11 20:24:14  1:45:25.659  600                0.377452         0.438357            0.880281        0.871662           2.02962          0.214676
    2020-07-11 20:25:32  1:46:43.684  700                0.376054         0.433869            0.881305        0.871345           2.03125          0.214808
    2020-07-11 20:26:52  1:48:02.985  800                0.374888         0.430476            0.882394        0.872722           2.03287          0.21252
    2020-07-11 20:28:14  1:49:24.995  900                0.373824         0.427652            0.883487        0.874197           2.03369          0.21114
    2020-07-11 20:29:37  1:50:47.865  1000               0.372784         0.42514             0.884641        0.875816           2.03369          0.211004
    2020-07-11 20:31:01  1:52:11.881  1100               0.371649         0.422628            0.885954        0.877147           2.03369          0.211176
    2020-07-11 20:32:26  1:53:37.643  1200               0.370498         0.42017             0.887282        0.877734           2.0345           0.209192
    2020-07-11 20:33:56  1:55:07.299  1300               0.369429         0.417946            0.888524        0.879669           2.0345           0.207552
    2020-07-11 20:35:00  1:56:11.105  1360               0.368809         0.416683            0.889246        0.88054            2.03369          0.206816

Variable Importances: 


variable    relative_importance    scaled_importance      percentage
----------  ---------------------  ---------------------  ----------------------
f10         7722454.0              1.0                    0.6846953073660439
e19         871827.1875            0.1128950962349533     0.07729874261671045
e18         240750.765625          0.03117542242724916    0.021345665441091625
f02         227124.78125           0.02941095942429699    0.020137545902949274
f27.F       186000.640625          0.02408569097659889    0.016491359586345927
---         ---                    ---                    ---
f09.K       8.973268508911133      1.161971118107163e-06  7.955961718628496e-07
c05.C       7.6248779296875        9.8736462913052e-07    6.760439282248813e-07
b03.N       6.3360137939453125     8.204663691030484e-07  5.617694727765137e-07
b02.N       3.8477745056152344     4.982580026524255e-07  3.4115491627369366e-07
a12.E       2.451904296875         3.175032569795819e-07  2.1739299012735485e-07


See the whole table with table.as_data_frame()


In [108]:
# Implementin the model to test data
grid_XGBoost = XGB_grid.get_grid(sort_by='accuracy', decreasing=True)
best_XG_Grid = grid_XGBoost.models[0]


preds = best_XG_Grid.predict(h2o_df_test)
df_test['Predicted'] = np.round(preds[2].as_data_frame(), 5)
df_preds_dt = df_test[['unique_id', 'Predicted']].copy()
df_test[['unique_id', 'Predicted']].to_csv(dirPOutput + '250k_XGBoost_Grid.csv', index=False)

# Kaggle Accuracy
# Accuracy = 0.86532
# From 0.8381

/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'e20' has levels not trained on: [30146, BE271, nan]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'e13' has levels not trained on: [Q, S]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'e24' has levels not trained on: [J, M, nan]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'e03' has levels not trained on: [J, nan]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'c09' has levels not trained on: [nan]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'a18' has levels not trained on: [D]
  warnings.warn(w)


### HyperOpt

In [15]:
# Splitting train data into valid data rather than n-folds because of speed efficiency. Even though I would prefer n-folds, I
# wanted to make this experiements since majority of people have used it in this way
train, valid = h2o_df_train.split_frame(ratios=[.7], seed = 2020)

In [16]:
# H20 Documentation says that the dependent variable has to be in factors
train[var_dep] = train[var_dep].asfactor()
valid[var_dep]   = valid[var_dep].asfactor()

In [27]:
# Objective
def hyperopt_objective(params):
    XGB_hpropt = H2OXGBoostEstimator(col_sample_rate_per_tree = .9, # Specify the column subsampling rate per tree. It
                                    # is multiplicative with col_sample_rate, so setting both parameters to 0.8, for example,
                                    # results in 64% of columns being considered at any given node to split.
                                    # 0.9 has been taken because it seemed the most common choice among the online blogs
                                   score_tree_interval = 100, # Scoring the model after every so many trees
                                   seed = 2020, max_runtime_secs= 1800, # Each XGBoost will run for maximum 30 minutes
                                   stopping_rounds = 4,  # Stops training when the option selected for stopping_metric doesn’t
                                    # improve for the specified number of  training rounds, based on a simple moving average. 
                                   stopping_metric = "AUC", # Since the kaggle score is based upon AUC, therefore I am using AUC as metrc
                                   stopping_tolerance = 1e-3, # to stop training if the improvement is less than this value. 1e-3 is the
                                     # default value
                                   ntrees = int(params['ntrees']), # Discussed under "Hyper-parameters to optimise"
                                   max_depth = int(params['max_depth']), # Discussed under "Hyper-parameters to optimise"
                                   learn_rate = params['learn_rate'], # Discussed under "Hyper-parameters to optimise"
                                   sample_rate = params['sample_rate'], # Discussed under "Hyper-parameters to optimise"
                                   min_rows = params['min_rows'] # Discussed under "Hyper-parameters to optimise"
                                   # ,max_runtime_secs = 18000, 
                                  )
    
    
    # training the model
    XGB_hpropt.train(x=vars_to_use,
               y='target',
               training_frame=train)
    

    # validating the model on validation frame
    preds = XGB_hpropt.predict(valid)
    preds = preds[2].as_data_frame()['p1'].values
    score = roc_auc_score(valid[var_dep].as_data_frame().as_matrix(), preds)
    
    return -score 

In [28]:
# Search Space
search_space = {
    'ntrees': hyperopt.hp.quniform('ntrees', 1200, 4000, 1)
    , 'max_depth': hyperopt.hp.quniform('max_depth', 4,8, 1)
    , 'learn_rate': hyperopt.hp.uniform('learn_rate', 0.001, 0.01)
    , 'sample_rate': hyperopt.hp.uniform('sample_rate', 0.4, 0.9)
    , 'min_rows': hyperopt.hp.quniform('min_rows', 10 , 40, 1)
}


In [29]:
# Optimisation
trials = hyperopt.Trials()

h2o.no_progress()
best = hyperopt.fmin(hyperopt_objective,
                     space=search_space,
                     algo=hyperopt.tpe.suggest,
                     max_evals=10, # Each model will take 30 minutes and there will be 10 models like this, thus 5 hours
                     trials=trials,
                     rstate=RandomState(2020)
)

print(best)

  0%|          | 0/10 [00:00<?, ?it/s, best loss: ?]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 10%|█         | 1/10 [30:57<4:38:35, 1857.28s/it, best loss: -0.8810321924378307]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 20%|██        | 2/10 [58:04<3:58:25, 1788.15s/it, best loss: -0.8811926515395831]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 30%|███       | 3/10 [1:04:12<2:38:54, 1362.09s/it, best loss: -0.8811926515395831]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 40%|████      | 4/10 [1:33:56<2:28:53, 1488.83s/it, best loss: -0.8820504208614022]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 50%|█████     | 5/10 [2:00:04<2:06:02, 1512.51s/it, best loss: -0.8820504208614022]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 60%|██████    | 6/10 [2:31:02<1:47:45, 1616.27s/it, best loss: -0.8820643226233149]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 70%|███████   | 7/10 [2:45:01<1:09:09, 1383.06s/it, best loss: -0.8820643226233149]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 80%|████████  | 8/10 [2:55:17<38:25, 1152.86s/it, best loss: -0.8820643226233149]  

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 90%|█████████ | 9/10 [3:26:04<22:41, 1361.22s/it, best loss: -0.8820643226233149]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



100%|██████████| 10/10 [3:40:01<00:00, 1203.76s/it, best loss: -0.8820643226233149]
{'learn_rate': 0.00632500998161564, 'max_depth': 8.0, 'min_rows': 37.0, 'ntrees': 3785.0, 'sample_rate': 0.7140605409926624}


In [25]:
# The best model returned the value of 
#     {'learn_rate': 0.00632500998161564,
#      'max_depth': 8.0,
#      'min_rows': 37.0,
#      'ntrees': 3785.0,
#      'sample_rate': 0.7140605409926624}

# I actually did not run the model next time but by mistake run this cell, therefore got this error
best

NameError: name 'best' is not defined

In [19]:
# Implementing the best model returned by Hyperopt
XGB_hpropt_bst = H2OXGBoostEstimator(col_sample_rate_per_tree = .9,
                               score_tree_interval = 100,
                               seed = 2020, max_runtime_secs= 3600,
                               stopping_rounds = 4, 
                               stopping_metric = "AUC", 
                               stopping_tolerance = 1e-3,
                               ntrees = 3785,
                               max_depth = 8,
                               learn_rate = 0.00632500998161564,
                               sample_rate = 0.7140605409926624,
                               min_rows = 37, nfolds = 5
                               # ,max_runtime_secs = 18000, 
                              )


XGB_hpropt_bst.train(x=vars_to_use,
           y='target',
           training_frame=train)

xgboost Model Build progress: |███████████████████████████████████████████| 100%


In [21]:
preds = XGB_hpropt_bst.predict(h2o_df_test)
df_test['Predicted'] = np.round(preds[2].as_data_frame(), 5)
df_preds_dt = df_test[['unique_id', 'Predicted']].copy()
df_test[['unique_id', 'Predicted']].to_csv(dirPOutput + '250k_XGBoost_hyperoptt_test_2ndver.csv', index=False)



xgboost prediction progress: |████████████████████████████████████████████| 100%


/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'e20' has levels not trained on: [30146, BE271, nan]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'e13' has levels not trained on: [Q, S]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'e24' has levels not trained on: [J, M, nan]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'e03' has levels not trained on: [J, nan]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'f03_e11' has levels not trained on: [C_G]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packages/h2o/job.py:69: UserWarning: Test/Validation dataset column 'c09' has levels not trained on: [nan]
  warnings.warn(w)
/opt/conda/lib/python3.6/site-packag

In [22]:
XGB_hpropt_bst

Model Details
H2OXGBoostEstimator :  XGBoost
Model Key:  XGBoost_model_python_1594806448431_1


ModelMetricsBinomial: xgboost
** Reported on train data. **

MSE: 0.13771069972813313
RMSE: 0.37109392305470745
LogLoss: 0.42207939087711466
Mean Per-Class Error: 0.20218874513834284
AUC: 0.8863087173166594
pr_auc: 0.8771276603825596
Gini: 0.7726174346333188
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.4171403104482695: 


0      1       Error    Rate
-----  -----  ------  -------  ------------------
0      63288  25722   0.289    (25722.0/89010.0)
1      10819  75329   0.1256   (10819.0/86148.0)
Total  74107  101051  0.2086   (36541.0/175158.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.41714      0.804801  231
max f2                       0.188044     0.881693  318
max f0point5                 0.627263     0.806768  150
max accuracy                 0.484833     0.797446  204
max precision                0.987655     1         0
max recall                   0.00628103   1         398
max specificity              0.987655     1         0
max absolute_mcc             0.483088     0.595915  205
max min_per_class_accuracy   0.507354     0.796281  195
max mean_per_class_accuracy  0.484833     0.797811  204

Gains/Lift Table: Avg response rate: 49.18 %, avg score: 49.19 %



group    cumulative_data_fraction    lower_threshold    lift       cumulative_lift    response_rate    score      cumulative_response_rate    cumulative_score    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  ---------  -----------------  ---------------  ---------  --------------------------  ------------------  --------------  -------------------------  --------  -----------------
    1        0.0100024                   0.977705           2.02162    2.02162            0.994292         0.982796   0.994292                    0.982796            0.020221        0.020221                   102.162   102.162
    2        0.0200048                   0.967772           2.01581    2.01872            0.991438         0.97249    0.992865                    0.977643            0.020163        0.040384                   101.581   101.872
    3        0.0300015                   0.960133           2.01348    2.01697            0.990291         0.963816   0.992008                    0.973036            0.0201282       0.0605121                  101.348   101.697
    4        0.0400039                   0.953601           2.01117    2.01552            0.989155         0.956836   0.991294                    0.968985            0.0201165       0.0806287                  101.117   101.552
    5        0.0500006                   0.947329           1.99723    2.01186            0.982296         0.950491   0.989495                    0.965288            0.0199656       0.100594                   99.7225   101.186
    6        0.100001                    0.917497           1.95359    1.98273            0.960836         0.93271    0.975166                    0.948999            0.0976807       0.198275                   95.3592   98.2728
    7        0.150002                    0.879781           1.87953    1.94833            0.924412         0.899677   0.958248                    0.932558            0.0939778       0.292253                   87.9535   94.833
    8        0.200002                    0.837241           1.78923    1.90855            0.879995         0.858935   0.938685                    0.914152            0.0894623       0.381715                   78.9226   90.8554
    9        0.300003                    0.737254           1.64854    1.82188            0.810802         0.788493   0.896057                    0.872266            0.164856        0.546571                   64.8539   82.1883
    10       0.399999                    0.612294           1.39069    1.71409            0.683985         0.676056   0.843041                    0.823216            0.139063        0.685634                   39.0694   71.409
    11       0.5                         0.501648           1.16403    1.60408            0.572505         0.554761   0.788933                    0.769524            0.116404        0.802038                   16.403    60.4077
    12       0.600001                    0.391488           0.906105   1.48775            0.44565          0.44732    0.731719                    0.715823            0.0906115       0.89265                    -9.38953  48.7747
    13       0.699997                    0.264305           0.612578   1.36273            0.301285         0.330108   0.670231                    0.660723            0.061255        0.953905                   -38.7422  36.2728
    14       0.799998                    0.136731           0.334653   1.23422            0.164592         0.197727   0.607025                    0.602848            0.0334657       0.987371                   -66.5347  23.4217
    15       0.899999                    0.0337288          0.113408   1.10968            0.0557776        0.0829282  0.545775                    0.545078            0.0113409       0.998712                   -88.6592  10.9681
    16       1                           0.00299243         0.0128847  1                  0.00633706       0.0131238  0.49183    



ModelMetricsBinomial: xgboost
** Reported on cross-validation data. **

MSE: 0.14274577098723212
RMSE: 0.37781711314766053
LogLoss: 0.4360396586479798
Mean Per-Class Error: 0.20995014317543115
AUC: 0.8774186628877368
pr_auc: 0.8667499376145853
Gini: 0.7548373257754737
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.4241522872324221: 


0      1       Error    Rate
-----  -----  ------  -------  ------------------
0      63112  25898   0.291    (25898.0/89010.0)
1      11793  74355   0.1369   (11793.0/86148.0)
Total  74905  100253  0.2152   (37691.0/175158.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.424152     0.797796  230
max f2                       0.197865     0.877526  316
max f0point5                 0.613705     0.79783   154
max accuracy                 0.484069     0.789676  206
max precision                0.988521     0.995169  0
max recall                   0.00624244   1         398
max specificity              0.988521     0.999989  0
max absolute_mcc             0.453039     0.580559  219
max min_per_class_accuracy   0.506348     0.788451  197
max mean_per_class_accuracy  0.48154      0.79005   207

Gains/Lift Table: Avg response rate: 49.18 %, avg score: 49.19 %



group    cumulative_data_fraction    lower_threshold    lift       cumulative_lift    response_rate    score      cumulative_response_rate    cumulative_score    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  ---------  -----------------  ---------------  ---------  --------------------------  ------------------  --------------  -------------------------  --------  -----------------
    1        0.0100024                   0.976273           2.01233    2.01233            0.989726         0.981676   0.989726                    0.981676            0.0201282       0.0201282                  101.233   101.233
    2        0.0200048                   0.966481           2.01001    2.01117            0.988584         0.971287   0.989155                    0.976481            0.0201049       0.0402331                  101.001   101.117
    3        0.0300015                   0.959067           1.99258    2.00498            0.980011         0.962593   0.986108                    0.971854            0.0199192       0.0601523                  99.2581   100.498
    4        0.0400039                   0.952429           1.9868     2.00043            0.977169         0.955646   0.983873                    0.967801            0.0198728       0.0800251                  98.6801   100.043
    5        0.0500006                   0.946416           1.97632    1.99561            0.972016         0.949366   0.981503                    0.964115            0.0197567       0.0997818                  97.6324   99.5613
    6        0.100001                    0.916508           1.93223    1.96392            0.950331         0.931595   0.965917                    0.947855            0.0966128       0.196395                   93.2234   96.3923
    7        0.150002                    0.878868           1.85028    1.92604            0.910025         0.89869    0.947286                    0.931467            0.0925152       0.28891                    85.0283   92.6043
    8        0.200002                    0.834736           1.76764    1.88644            0.869377         0.857007   0.927809                    0.912852            0.0883828       0.377293                   76.7635   88.6441
    9        0.300003                    0.734672           1.62544    1.79944            0.799441         0.786148   0.885019                    0.870617            0.162546        0.539838                   62.544    79.9441
    10       0.399999                    0.60976            1.38628    1.69616            0.681816         0.674178   0.834221                    0.821509            0.138622        0.67846                    38.6282   69.6156
    11       0.5                         0.50086            1.1566     1.58824            0.568851         0.553377   0.781146                    0.767882            0.115661        0.794122                   15.6601   58.8243
    12       0.600001                    0.394143           0.90831    1.47492            0.446734         0.447904   0.72541                     0.714552            0.0908321       0.884954                   -9.16898  47.492
    13       0.699997                    0.266136           0.634054   1.3548             0.311847         0.332563   0.666332                    0.659984            0.0634025       0.948356                   -36.5946  35.4801
    14       0.799998                    0.139373           0.359842   1.23043            0.176981         0.200237   0.605162                    0.602515            0.0359846       0.984341                   -64.0158  23.043
    15       0.899999                    0.0354208          0.138249   1.10907            0.067995         0.0849552  0.545476                    0.545008            0.013825        0.998166                   -86.1751  10.9075
    16       1                           0.00200994         0.0183403  1                  0.00902032       0.014055   0.49183    


Cross-Validation Metrics Summary: 


mean      sd           cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid
-----------------------  --------  -----------  ------------  ------------  ------------  ------------  ------------
accuracy                 0.783066  0.00325388   0.786448      0.783966      0.774331      0.783369      0.787218
auc                      0.877469  0.000902725  0.878085      0.87721       0.875248      0.877683      0.879119
err                      0.216934  0.00325388   0.213552      0.216034      0.225669      0.216631      0.212782
err_count                7600      131.761      7507          7580          7958          7534          7421
f0point5                 0.759555  0.00502527   0.763053      0.762184      0.7456        0.761467      0.765471
f1                       0.798307  0.00100964   0.798756      0.799471      0.795624      0.7995        0.798183
f2                       0.841349  0.00448211   0.837964      0.840593      0.852842      0.841531      0.833816
lift_top_group           2.01004   0.00422761   2.0151        2.00398       2.01622       2.00175       2.01314
logloss                  0.436032  0.00181538   0.435669      0.436576      0.439769      0.436404      0.43174
max_per_class_error      0.3038    0.0147422    0.290042      0.300711      0.342533      0.303691      0.282022
mcc                      0.577372  0.00364615   0.582198      0.577632      0.568071      0.576816      0.582143
mean_per_class_accuracy  0.784504  0.00304825   0.788136      0.784866      0.776629      0.784203      0.788685
mean_per_class_error     0.215496  0.00304825   0.211864      0.215134      0.223371      0.215797      0.211315
mse                      0.142743  0.00057792   0.142512      0.142864      0.144138      0.142599      0.141601
precision                0.73579   0.00733467   0.740973      0.739201      0.715606      0.73806       0.745113
r2                       0.428861  0.00230553   0.429686      0.428481      0.423233      0.429551      0.433353
recall                   0.872807  0.00869714   0.866314      0.870442      0.89579       0.872097      0.859393
rmse                     0.377812  0.000764345  0.377508      0.377974      0.379655      0.377623      0.376299
specificity              0.6962    0.0147422    0.709958      0.699289      0.657467      0.696309      0.717978

Scoring History: 


timestamp            duration           number_of_trees    training_rmse    training_logloss    training_auc    training_pr_auc    training_lift    training_classification_error
--  -------------------  -----------------  -----------------  ---------------  ------------------  --------------  -----------------  ---------------  -------------------------------
    2020-07-15 10:45:12  49 min 10.086 sec  0                  0.5              0.693147            0.5             0                  1                0.50817
    2020-07-15 10:46:06  50 min  4.280 sec  100                0.427024         0.552687            0.87144         0.861688           2.01001          0.224768
    2020-07-15 10:46:59  50 min 57.128 sec  200                0.397249         0.4913              0.874531        0.865515           2.01581          0.220076
    2020-07-15 10:47:54  51 min 52.382 sec  300                0.385695         0.462719            0.87644         0.866953           2.0193           0.216416
    2020-07-15 10:48:47  52 min 45.473 sec  400                0.380911         0.448562            0.877745        0.867998           2.0193           0.216964
    2020-07-15 10:49:41  53 min 38.838 sec  500                0.378461         0.44065             0.878901        0.869782           2.02046          0.215725
    2020-07-15 10:50:35  54 min 33.051 sec  600                0.37681          0.435558            0.880116        0.870925           2.02046          0.214007
    2020-07-15 10:51:30  55 min 27.952 sec  700                0.375516         0.431955            0.881319        0.871063           2.02046          0.21315
    2020-07-15 10:52:26  56 min 23.782 sec  800                0.374374         0.429146            0.882544        0.872799           2.0193           0.212505
    2020-07-15 10:53:25  57 min 23.477 sec  900                0.373108         0.426314            0.883973        0.874567           2.01814          0.211341
    2020-07-15 10:54:24  58 min 22.320 sec  1000               0.371804         0.423558            0.885487        0.876307           2.0193           0.21033
    2020-07-15 10:55:06  59 min  3.843 sec  1062               0.371094         0.422079            0.886309        0.877128           2.02162          0.208617

Variable Importances: 


variable    relative_importance    scaled_importance       percentage
----------  ---------------------  ----------------------  ----------------------
f10         4053835.75             1.0                     0.6988022612978955
e19         446379.34375           0.1101128341842661      0.07694709752588536
e18         121911.71875           0.030073176682109036    0.021015203869868937
f02         121454.71875           0.029960443944972364    0.020936425978235525
f27.F       93993.90625            0.023186412091313762    0.016202717200794924
---         ---                    ---                     ---
e21.P       10.039320945739746     2.47649918863628e-06    1.7305832331214358e-06
e21.J       8.786211013793945      2.167382093315928e-06   1.5145715079057367e-06
c05.A       8.73931884765625       2.155814736119057e-06   1.506488212539323e-06
e21.N       5.731616497039795      1.4138748707418141e-06  9.880189568666492e-07
c07.A       1.13916015625          2.810079703525235e-07   1.963690051250754e-07


See the whole table with table.as_data_frame()


##### Old model (not so important)
I ran 40 models for 2 mins each, and was not expecting a better results since each model got only a time frame of 2 mins. But to my surprise this model yielded better Kaggle score of 0.867 compare to 0.865 of the above hyperopt

In [86]:
# Optimisation - OLD
trials = hyperopt.Trials()

h2o.no_progress()
best = hyperopt.fmin(hyperopt_objective,
                     space=search_space,
                     algo=hyperopt.tpe.suggest,
                     max_evals=40,
                     trials=trials,
                     rstate=RandomState(2020)
)

print(best)

  0%|          | 0/40 [00:00<?, ?it/s, best loss: ?]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



  2%|▎         | 1/40 [02:04<1:20:39, 124.10s/it, best loss: -0.8674585083914759]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



  5%|▌         | 2/40 [04:08<1:18:42, 124.28s/it, best loss: -0.870175121415311] 

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



  8%|▊         | 3/40 [06:14<1:16:54, 124.73s/it, best loss: -0.870175121415311]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 10%|█         | 4/40 [08:18<1:14:44, 124.56s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 12%|█▎        | 5/40 [10:24<1:12:52, 124.93s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 15%|█▌        | 6/40 [12:29<1:10:45, 124.87s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 18%|█▊        | 7/40 [14:33<1:08:31, 124.58s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 20%|██        | 8/40 [16:38<1:06:37, 124.93s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 22%|██▎       | 9/40 [18:43<1:04:25, 124.69s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 25%|██▌       | 10/40 [20:47<1:02:16, 124.55s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 28%|██▊       | 11/40 [22:51<1:00:13, 124.60s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 30%|███       | 12/40 [24:57<58:12, 124.72s/it, best loss: -0.871667523546091]  

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 32%|███▎      | 13/40 [27:01<56:02, 124.55s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 35%|███▌      | 14/40 [29:05<53:54, 124.42s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 38%|███▊      | 15/40 [31:09<51:46, 124.27s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 40%|████      | 16/40 [33:13<49:39, 124.15s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 42%|████▎     | 17/40 [35:17<47:37, 124.24s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 45%|████▌     | 18/40 [37:21<45:31, 124.15s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 48%|████▊     | 19/40 [39:25<43:26, 124.10s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 50%|█████     | 20/40 [41:29<41:22, 124.13s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 52%|█████▎    | 21/40 [43:34<39:22, 124.32s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 55%|█████▌    | 22/40 [45:39<37:20, 124.46s/it, best loss: -0.871667523546091]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 57%|█████▊    | 23/40 [47:43<35:17, 124.56s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 60%|██████    | 24/40 [49:48<33:10, 124.44s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 62%|██████▎   | 25/40 [51:52<31:08, 124.54s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 65%|██████▌   | 26/40 [53:57<29:01, 124.42s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 68%|██████▊   | 27/40 [56:01<26:59, 124.56s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 70%|███████   | 28/40 [58:06<24:55, 124.61s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 72%|███████▎  | 29/40 [1:00:11<22:51, 124.66s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 75%|███████▌  | 30/40 [1:02:15<20:45, 124.53s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 78%|███████▊  | 31/40 [1:04:20<18:41, 124.61s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 80%|████████  | 32/40 [1:06:25<16:36, 124.59s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 82%|████████▎ | 33/40 [1:08:29<14:32, 124.64s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 85%|████████▌ | 34/40 [1:10:34<12:27, 124.64s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 88%|████████▊ | 35/40 [1:12:38<10:22, 124.59s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 90%|█████████ | 36/40 [1:14:43<08:17, 124.47s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 92%|█████████▎| 37/40 [1:16:47<06:13, 124.55s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 95%|█████████▌| 38/40 [1:18:51<04:08, 124.42s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



 98%|█████████▊| 39/40 [1:20:56<02:04, 124.37s/it, best loss: -0.8718621667871277]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:26: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.



100%|██████████| 40/40 [1:23:00<00:00, 124.49s/it, best loss: -0.8719438506668313]
{'learn_rate': 0.008933646880323847, 'max_depth': 8.0, 'min_rows': 10.0, 'ntrees': 2908.0, 'sample_rate': 0.5396549266120586}


In [109]:
best.

{'learn_rate': 0.008933646880323847,
 'max_depth': 8.0,
 'min_rows': 10.0,
 'ntrees': 2908.0,
 'sample_rate': 0.5396549266120586}